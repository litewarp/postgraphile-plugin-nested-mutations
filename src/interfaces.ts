import type {
  PgCodec,
  PgCodecAttribute,
  PgCodecAttributes,
  PgDeleteSingleStep,
  PgInsertSingleStep,
  PgResourceUnique,
  PgUpdateSingleStep,
} from '@dataplan/pg';
import type { PgTableResource } from '@graphile-contrib/pg-many-to-many';
import type {} from 'postgraphile';
import type { pgNestedMutationFields } from './get-nested-relationships';

export type PgNestedConnectorsInflectionFn = (
  this: GraphileBuild.Inflection,
  details: Omit<PgNestedMutationRelationship, 'mutationFields'>,
) => string;

export type PgTableRelationship = ReturnType<
  PgTableResource['getRelations']
>[1];

export type BuildInputObjectArguments = Parameters<
  GraphileBuild.Build['registerInputObjectType']
>;

export interface PgNestedUniqueAttributeCodecs<
  TAttributes extends PgCodecAttributes = PgCodecAttributes,
> extends PgResourceUnique<TAttributes> {
  isNodeId?: boolean;
  codecs: Record<string, PgCodecAttribute>;
}

export type PgNestedDataPlanStep =
  | PgInsertSingleStep
  | PgUpdateSingleStep
  | PgDeleteSingleStep;

export type PgNestedMutationFieldNames =
  (typeof pgNestedMutationFields)[number];

export interface PgNestedMutationFieldDetails {
  typeName: string;
  fieldName: string;
}

export interface PgNestedTableMutationFields {
  input: PgNestedMutationFieldDetails;
  create?: PgNestedMutationFieldDetails;
  connectByKeys?: PgNestedMutationFieldDetails[];
  connectByNodeId?: PgNestedMutationFieldDetails;
  updateByKeys?: PgNestedMutationFieldDetails[];
  updateByNodeId?: PgNestedMutationFieldDetails;
}

export interface PgNestedMutationRelationship {
  leftTable: PgTableResource;
  rightTable: PgTableResource;
  relationName: string;
  isReverse?: boolean;
  isUnique?: boolean;
  localAttributes: readonly string[];
  remoteAttributes: readonly string[];
  localUnique: PgNestedUniqueAttributeCodecs;
  remoteUnique: PgNestedUniqueAttributeCodecs;
  mutationFields: PgNestedTableMutationFields;
  tableFieldName: string;
}

declare global {
  namespace GraphileBuild {
    interface BehaviorEntities {
      pgNestedCreate: PgNestedMutationRelationship;
    }

    interface Build {
      /**
       *
       */
      pgNestedMutationRelationships: Map<
        PgCodec,
        PgNestedMutationRelationship[]
      >;

      /**
       * Store of types created to deduplicate
       */
      pgNestedMutationInputTypes: Set<string>;

      // returns the field in the input object
      // should be tablename for insert
      // patch for patch
      pgNestedMutationInputObjMap: Map<
        string,
        { field: string; codec: PgCodec }
      >;

      //
    }
    interface Inflection {
      /**
       * Create Fields
       */
      nestedCreateFieldName: PgNestedConnectorsInflectionFn;
      nestedCreateInputType: PgNestedConnectorsInflectionFn;

      /**
       * Connect Fields
       */
      nestedConnectByNodeIdInputType: PgNestedConnectorsInflectionFn;
      nestedConnectByNodeIdFieldName: PgNestedConnectorsInflectionFn;
      nestedConnectByKeyInputType: PgNestedConnectorsInflectionFn;
      nestedConnectByKeyFieldName: PgNestedConnectorsInflectionFn;
      /**
       * Update Fields
       */
      nestedUpdateByNodeIdInputType: PgNestedConnectorsInflectionFn;
      nestedUpdateByNodeIdFieldName: PgNestedConnectorsInflectionFn;
      nestedUpdateByKeyInputType: PgNestedConnectorsInflectionFn;
      nestedUpdateByKeyFieldName: PgNestedConnectorsInflectionFn;
      nestedUpdatePatchType: PgNestedConnectorsInflectionFn;

      /**
       * Input type and field for the connector
       */
      nestedConnectorFieldType: PgNestedConnectorsInflectionFn;
      nestedConnectorFieldName: PgNestedConnectorsInflectionFn;
    }

    interface ScopeInputObject {
      /**
       * General Flag for all nested mutation input types
       */
      isNestedMutationInputType?: boolean;
      /**
       * Indicates that the type is for the create field
       */
      isNestedMutationCreateInputType?: boolean;
      /**
       * Indicates that the type is for table to codec
       */
      isNestedInverseMutation?: boolean;
      /**
       * connect_by_key or connect_by_key_a_and_key_b field
       */
      isNestedMutationConnectByKeyType?: boolean;
      /**
       * connect_by_node_id field
       */
      isNestedMutationConnectByNodeIdType?: boolean;
      /**
       * type for the relationship field on the parent input
       * contains all the nested mutation input types
       */
      isNestedMutationConnectorType?: boolean;
      /**
       * delete_by_key or delete_by_key_a_and_key_b field
       */
      isNestedMutationDeleteByKeyType?: boolean;
      /**
       * delete_by_node_id field
       */
      isNestedMutationDeleteByNodeInputType?: boolean;
      /**
       * update_by_key or update_by_key_a_and_key_b field
       */
      isNestedMutationUpdateByKeyType?: boolean;
      /**
       * update_by_node_id field
       */
      isNestedMutationUpdateByNodeIdType?: boolean;
      /**
       * patch types for the update field for the foreign input
       */
      isNestedMutationPatchType?: boolean;
      /**
       * field which resolves to the the connector type
       */
      isNestedMutationInputField?: boolean;
    }

    interface ScopeInputObjectFieldsField {}
  }
}
