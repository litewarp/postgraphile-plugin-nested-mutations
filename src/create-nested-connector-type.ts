import { GrafastInputFieldConfigMap } from 'grafast';
import type {
  PgNestedMutationRelationDetails,
  ResourceNestedMutationDetails,
} from './interfaces';

// need to define the Input Types for
// 1. connect
// 2. update
// 3. delete

type BuildInputObjectArguments = Parameters<
  GraphileBuild.Build['registerInputObjectType']
>;

export function createNestedConnectorType(
  details: PgNestedMutationRelationDetails,
  build: GraphileBuild.Build,
) {
  const {
    inflection,
    graphql: { GraphQLNonNull, GraphQLInputObjectType, GraphQLBoolean },
    grafast: {},
    getTypeByName,
    options: {},
    nullableIf,
  } = build;

  const { name, table, relationship } = details;

  const tableName = build.inflection.tableFieldName(table);

  const foreignTable = relationship.remoteResource;
  const foreignTableName = build.inflection.tableFieldName(foreignTable);

  const connectable = Boolean(foreignTable.uniques.length);

  // to do
  // const createable = true;
  // const updateable = true;
  // const deleteable = true;
  // return early if not connectable, createable, updateable, deletable or omit on read for the foreign table
  // if (
  //   (!connectable && !creatable && !deleteable && !updateable) ||
  //   omit(foreignTable, 'read')
  // ) {
  //   return;
  // }

  /* start from most granular and work out */
  /* connector fields first */
  /* deleter fields second */
  /* updater fields third */
  /* create and field types last */

  const connectByKeyInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const connectByNodeIdInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const deleteByKeyInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const deleteByNodeIdInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const updateByKeyInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const updateByNodeIdInputType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const updateByPatchType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const nestedConnectorFieldType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const nestedConnectorCreateType: BuildInputObjectArguments = [
    'typename',
    {},
    () => ({
      description: '',
      fields: ({ fieldWithHooks }) => {
        return {};
      },
    }),
    'info',
  ];

  const nestedTypes = [
    connectByKeyInputType,
    connectByNodeIdInputType,
    deleteByKeyInputType,
    deleteByNodeIdInputType,
    updateByKeyInputType,
    updateByNodeIdInputType,
    updateByPatchType,
    nestedConnectorFieldType,
    nestedConnectorCreateType,
  ];

  for (const type of nestedTypes) {
    build.registerInputObjectType(...type);
  }
}

function pgNestedFieldName(
  rel: ResourceNestedMutationDetails,
  inflection: GraphileBuild.Build['inflection'],
): string {
  const {
    isForward,
    table,
    foreignTable,
    leftTableKeyNames,
    foreignTableKeyNames,
  } = rel;

  const tableFieldName = inflection.tableFieldName(foreignTable);

  // filter out non applicable constraints
  // dont include constraints where read is omitted

  // multiple fk check to see if it's a many to many relationship
  const multipleFKs = false;

  const isUnique = foreignTable.isUnique;

  const computedReverseMutationName = inflection.camelCase(
    isUnique ? tableFieldName : inflection.pluralize(tableFieldName),
  );

  if (isForward) {
    return inflection.camelCase(
      `${tableFieldName}_to_${leftTableKeyNames.join('_and_')}`,
    );
  }

  if (!multipleFKs) {
    return inflection.camelCase(
      `${computedReverseMutationName}_using_${foreignTableKeyNames.join(
        '_and_',
      )}`,
    );
  }

  // tables have mutliple relations between them
  return inflection.camelCase(
    `${computedReverseMutationName}_to_${leftTableKeyNames.join(
      '_and_',
    )}_using_${foreignTableKeyNames.join('_and_')}`,
  );
}
