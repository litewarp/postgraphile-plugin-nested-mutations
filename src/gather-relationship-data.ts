import { PgNestedMutationRelationDetails, PgTableResource } from './interfaces';

/**
 * For each table, gather all the relationships that are nested
 * and prepare them for the nested connector plugin.
 */
export function gatherRelationshipData(
  leftTable: PgTableResource,
  build: GraphileBuild.Build,
): PgNestedMutationRelationDetails[] {
  const relations = leftTable.getRelations();

  return Object.entries(relations).reduce(
    (acc, [relationName, rightRelation]) => {
      const details: PgNestedMutationRelationDetails = {
        relationName: relationName,
        table: leftTable,
        relationship: rightRelation,
        typeName: '',
      };

      return [...acc, details];
    },
    [] as PgNestedMutationRelationDetails[],
  );
}
