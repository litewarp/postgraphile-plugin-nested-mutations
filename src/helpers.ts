import type { PgResource } from '@dataplan/pg';
import { PgInsertSingleStep, PgUpdateSingleStep } from '@dataplan/pg';
import type { PgTableResource } from '@graphile-contrib/pg-many-to-many';
import type { ExecutableStep } from 'grafast';

export function isPgTableResource(r: PgResource): r is PgTableResource {
  return Boolean(r.codec.attributes) && !r.parameters;
}

export const isInsertable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  return build.behavior.pgResourceMatches(resource, 'resource:insert') === true;
};

export const isUpdatable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return Boolean(build.behavior.pgResourceMatches(resource, 'resource:update'));
};

export const isDeletable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return Boolean(build.behavior.pgResourceMatches(resource, 'resource:delete'));
};

export const getCRUDBehavior = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return {};
  if (!resource.codec.attributes) return {};
  if (resource.codec.polymorphism) return {};
  if (resource.codec.isAnonymous) return {};
  if (!resource.uniques || resource.uniques.length < 1) return {};

  return {
    insertable: isInsertable(build, resource),
    // connectable: isInsertable(build, resource), // TODO
    updatable: isUpdatable(build, resource),
    deletable: isDeletable(build, resource),
  };
};

export function isInsertOrUpdate(
  $step: ExecutableStep,
): $step is PgInsertSingleStep | PgUpdateSingleStep {
  return (
    $step instanceof PgInsertSingleStep || $step instanceof PgUpdateSingleStep
  );
}

export let inspect: (
  obj: any,
  options?: { colors?: boolean; depth?: number },
) => string;

try {
  inspect = require('node:util').inspect;
  if (typeof inspect !== 'function') {
    throw new Error('Failed to load inspect');
  }
} catch {
  inspect = (obj) => {
    return Array.isArray(obj) ||
      !obj ||
      Object.getPrototypeOf(obj) === null ||
      Object.getPrototypeOf(obj) === Object.prototype
      ? JSON.stringify(obj)
      : String(obj);
  };
}
