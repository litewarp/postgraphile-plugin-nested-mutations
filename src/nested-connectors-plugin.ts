import { GraphileConfig } from 'graphile-build';
import { PgResource } from '@dataplan/pg';

declare global {
  namespace GraphileBuild {
    interface Inflection {
      nestedConnectByNodeIdField(
        this: Inflection,
        resource: PgResource,
      ): string;
    }
  }
}

const config: GraphileConfig.Plugin = {
  name: '',
  version: '',
  description: '',
  experimental: true,
  provides: ['Nested Mutations'],
  after: [''],
  before: [''],
  inflection: {
    add: {},
  },
  gather: {},
  schema: {},
};
