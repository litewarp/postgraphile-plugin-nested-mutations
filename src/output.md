# Generate Types for Nested Mutations

## Input

### Nested Connector Type

Description: Input for the nested mutation of `foreignTableName` in the `tableTypeName` mutation

Typename Examples
- OpportunityIssuerIdFkeyInput
  - connectByOrganizationId
  - deleteByShortUrl
  - updateById
  - create

### Nested Connection Create Input

Description: The `foreignTableName` to be created by this mutation

Typename Examples

- OpportunityIsserIdFkeyOrganizationCreateInput
  - all organization input fields; and
  - all nested mutation fields

## Connect

### Nested Connect By Key Input Type

Description: The fields on `tableFieldName` to look up the row to connect

Typename Examples
- OrganizationOrganizationPkeyConnect
  - organizationId

- OrganizationOrganizationShortUrlKeyConnect
  - shortUrl

- UserOrganizationFollowUserOrganizationFollowPkeyConnect
  - userId
  - organizationId

### Nested Connect By Node Id Input Type 

Description: The globally unique `ID` lookup for the row to connect

Typename Examples
- OrganizationNodeIdConnect
- UserNodeIdConnect
  - id: ID!

## Delete

### Nested Delete By Key Input Type

### Nested Delete By Node Id Input Type

## Update

### Nested Update By Key Input Type

### Nested Update By Node Id Input Type

### Nested Update By Patch Type