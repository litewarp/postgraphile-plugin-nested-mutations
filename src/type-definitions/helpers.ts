import { PgResource } from '@dataplan/pg';
import { PgTableResource } from '@graphile-contrib/pg-many-to-many';

export function isPgTableResource(r: PgResource): r is PgTableResource {
  return Boolean(r.codec.attributes) && !r.parameters;
}

export const isInsertable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  return build.behavior.pgResourceMatches(resource, 'resource:insert') === true;
};

export const isUpdatable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return !!build.behavior.pgResourceMatches(resource, 'resource:update');
};

export const isDeletable = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return !!build.behavior.pgResourceMatches(resource, 'resource:delete');
};

export const getCRUDBehavior = (
  build: GraphileBuild.Build,
  resource: PgResource<any, any, any, any, any>,
) => {
  if (resource.parameters) return {};
  if (!resource.codec.attributes) return {};
  if (resource.codec.polymorphism) return {};
  if (resource.codec.isAnonymous) return {};
  if (!resource.uniques || resource.uniques.length < 1) return {};

  return {
    insertable: isInsertable(build, resource),
    // connectable: isInsertable(build, resource), // TODO
    updatable: isUpdatable(build, resource),
    deletable: isDeletable(build, resource),
  };
};
