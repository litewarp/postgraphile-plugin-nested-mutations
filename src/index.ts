import type { GraphileConfig } from 'graphile-build';
// import { PostGraphileNestedConnectorsPlugin } from './type-definitions/connect-plugin';
import { PostGraphileNestedTypesPlugin } from './type-definitions/type-deux-plugin';

export const NestedMutationPreset: GraphileConfig.Preset = {
  plugins: [PostGraphileNestedTypesPlugin],
};
