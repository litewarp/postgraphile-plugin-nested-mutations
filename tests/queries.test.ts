import { readFile, writeFile } from 'node:fs/promises';
import { existsSync, readdirSync } from 'node:fs';
import path from 'node:path';
import type { SchemaResult } from 'graphile-build';
import { makeSchema } from 'graphile-build';
import { PostGraphileAmberPreset } from 'postgraphile/presets/amber';
import { PgManyToManyPreset } from '@graphile-contrib/pg-many-to-many';
import { hookArgs, execute } from 'grafast';
import type { ExecutionArgs } from 'graphql';
import { parse, validate } from 'graphql';
import { PgSimplifyInflectionPreset } from '@graphile/simplify-inflection';
import type { Pool } from 'pg';
import { makeWithPgClientViaPgClientAlreadyInTransaction } from 'postgraphile/adaptors/pg';
import { NestedMutationPreset } from '../src';
import { withPgClient, withPgPool } from './helpers';
import { printOrderedSchema } from './print-ordered-schema';

const readFixtureForSqlSchema = async (sqlSchema: string, fixture: string) =>
  readFile(
    path.resolve(
      __dirname,
      'schemas',
      sqlSchema,
      'fixtures',
      'queries',
      fixture,
    ),
    'utf8',
  );

const createPostGraphileSchema = async (pgPool: Pool, sqlSchema: string) => {
  const gs = await makeSchema({
    extends: [
      PostGraphileAmberPreset,
      PgSimplifyInflectionPreset,
      PgManyToManyPreset,
      NestedMutationPreset,
    ],
    pgServices: [
      {
        name: 'main',
        adaptor: '@dataplan/pg/adaptors/pg',
        withPgClientKey: 'withPgClient',
        pgSettingsKey: 'pgSettings',
        pgSettingsForIntrospection: {},
        schemas: [sqlSchema],
        adaptorSettings: {
          pool: pgPool,
        },
      },
    ],
  });
  await writeFile(
    `./tmp/${sqlSchema}.graphql`,
    printOrderedSchema(gs.schema),
    'utf8',
  );
  return gs;
};

const getFixturesForSqlSchema = (sqlSchema: string) =>
  existsSync(
    path.resolve(__dirname, 'schemas', sqlSchema, 'fixtures', 'queries'),
  )
    ? readdirSync(
        path.resolve(__dirname, 'schemas', sqlSchema, 'fixtures', 'queries'),
      ).sort()
    : [];

const getSqlSchemas = () =>
  readdirSync(path.resolve(__dirname, 'schemas')).sort();

const sqlSchemas = getSqlSchemas();

let gqlSchema: SchemaResult;

beforeAll(async () => {
  // Ensure process.env.TEST_DATABASE_URL is set
  if (!process.env.TEST_DATABASE_URL) {
    console.error(
      'ERROR: No test database configured; aborting. To resolve this, ensure environmental variable TEST_DATABASE_URL is set.',
    );
    process.exit(1);
  }
});

describe.each(sqlSchemas)('%s', (sqlSchema) => {
  beforeEach(async () => {
    // reset db
    await withPgClient(async (pgClient) => {
      const schema = await readFile(
        path.resolve(__dirname, 'schemas', sqlSchema, 'schema.sql'),
        'utf8',
      );
      await pgClient.query(schema);
      const data = await readFile(
        path.resolve(__dirname, 'schemas', sqlSchema, 'data.sql'),
        'utf8',
      );
      await pgClient.query(data);
    });

    gqlSchema = await withPgPool(async (pool) =>
      createPostGraphileSchema(pool, sqlSchema),
    );
  });
  const fixtures = getFixturesForSqlSchema(sqlSchema);
  if (fixtures.length > 0) {
    test.each(fixtures)('query=%s', async (fixture) => {
      const { schema, resolvedPreset } = gqlSchema;
      const query = await readFixtureForSqlSchema(sqlSchema, fixture);
      const document = parse(query);
      const errors = validate(schema, document);
      if (errors.length > 0) {
        throw new Error(
          `GraphQL validation errors:\n${errors
            .map((e) => e.message)
            .join('\n')}`,
        );
      }
      const args: ExecutionArgs = {
        schema,
        document,
      };
      await hookArgs(args, resolvedPreset, {});
      const result = await withPgClient(async (pgClient) => {
        // We must override the context because we didn't use a pool above and so
        // we need to add our own client
        // NOTE: the withPgClient needed on context is **VERY DIFFERENT** to our
        // withPgClient test helper. We should rename our test helper ;)

        const contextWithPgClient =
          makeWithPgClientViaPgClientAlreadyInTransaction(pgClient, false);

        try {
          args.contextValue = {
            pgSettings: (args.contextValue as any).pgSettings,
            withPgClient: contextWithPgClient,
          };
          return (await execute(args)) as any;
        } finally {
          contextWithPgClient.release?.();
        }
      });
      if (result.errors) {
        console.log(result.errors.map((e: any) => e.originalError ?? e));
      }
      console.log(JSON.stringify(result.data, null, 2));
      expect(result).toMatchSnapshot();
    });
  }
});
