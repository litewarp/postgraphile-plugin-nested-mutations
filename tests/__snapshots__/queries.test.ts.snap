// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a query=inverse-create.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "dadParent": {
          "id": "WyJQYXJlbnQiLDNd",
        },
        "id": "WyJDaGlsZCIsMl0=",
        "momParent": null,
        "name": "Me first but what about required constraints",
      },
    },
  },
}
`;

exports[`a query=test.graphql 1`] = `
{
  "data": {
    "createParent": {
      "parent": {
        "childrenByDadParentId": {
          "nodes": [
            {
              "dadParent": {
                "id": "WyJQYXJlbnQiLDNd",
              },
              "id": "WyJDaGlsZCIsMl0=",
              "momParent": null,
              "name": "child 1 of test f1",
            },
          ],
        },
        "childrenByMomParentId": {
          "nodes": [
            {
              "id": "WyJDaGlsZCIsM10=",
              "momParent": {
                "id": "WyJQYXJlbnQiLDNd",
              },
              "name": "child 2 of test f1",
            },
          ],
        },
        "id": "WyJQYXJlbnQiLDNd",
        "name": "test f1",
      },
    },
  },
}
`;

exports[`b query=test.graphql 1`] = `
{
  "data": {
    "createJob": {
      "job": {
        "id": "WyJKb2IiLDFd",
        "jobRelationshipsByFromJobId": {
          "nodes": [],
        },
        "jobRelationshipsByToJobId": {
          "nodes": [
            {
              "fromJobId": null,
              "toJobId": 1,
              "type": "test",
            },
          ],
        },
      },
    },
  },
}
`;
